PolyEngine/Core/Src/pe/core/utils/Result.hpp

```diff
@@ -0,0 +1,280 @@
+
+		// @returns stored error message
+		constexpr const char* get() && { return m_value; }
+
+	private:
+		const char* m_value = DEFAULT_ERROR_MESSAGE;
+	};
+
+	Err()->Err<const char*>;
+
+	//------------------------------------------------------------------------------
+	// Result is an utility class that can be used when function can
+	// return result or error code. Its main purpose is to enable 
+	// writing exception free code. You can't call getError when Result
+	// has valid success value and you can't call getValue when Result
+	// has error value but for 
+	// @tparam T - type of correct value (value returned by method if 
+	//		its execution was successful
+	// @tparam E - type of error
+	template <typename T, typename E = const char*>
+	class Result final
+	{
+	public:
+		// You can't create empty Result. Result must always store valid 
+		// value or error value 
+		Result() = delete;
+
+		// We're ok with default implementations. Some of these ctors will be
+		// implicitly deleted if T or E aren't copy constructible or copyable
+		// and so on.
+		constexpr Result(Result&) = default;
+		constexpr Result(Result&&) = default;
+		constexpr Result& operator=(Result&) = default;
+		constexpr Result& operator=(Result&&) = default;
+
+		// Ctor taking ok value. Object created with this ctor will return 
+		// true from isOk and false from isErr.
+		// @param okVal - value for success result
+		constexpr Result(Ok<T> okVal) : m_value(std::move(okVal).get()) {}
+
+		// Ctor taking error value. Object created with this ctor will return
+		// false from isOk and true from isErr.
+		// @param errVal - value for error/invalid result
+		constexpr Result(Err<E> errVal) : m_value(std::move(errVal).get()) {}
+
+		// Ctor taking ok value. Object created with this ctor will return 
+		// true from isOk and false from isErr. Using this ctor rather than
+		// the one taing Ok<T> will save you one move operation.
+		// @param okVal - value for success result
+		explicit constexpr Result(T okVal) : m_value(std::move(okVal)) {}
+
+		// Ctor taking error value. Object created with this ctor will return
+		// false from isOk and true from isErr. Using this ctor rather than
+		// the one taing Err<T> will save you one move operation.
+		// @param errVal - value for error/invalid result
+		explicit constexpr Result(E errVal) : m_value(std::move(errVal)) {}
+		
+		// @returns true if result stores valid value, false if stores error
+		//		value
+		constexpr bool isOk() const { return m_value.index() == 0; }
+
+		// @returns false if result stores valid value, true if stores error
+		//		value
+		constexpr bool isErr() const { return !isOk(); }
+		
+		// @returns valid value if has valid value, crashes otherwise :(
+		constexpr T& value() { return std::get<T>(m_value); }
+
+		// @returns valid value if has valid value, crashes otherwise :(
+		constexpr const T& value() const { return std::get<T>(m_value); }
```

Comment:
Create change value() to be similar in nature to valueOr() (return T by val, overloaded for l and r-value references).
Add valueRef() which returns const T&.